src/components/multi-select/multi-select.test.jsx

```diff
@@ -0,0 +1,500 @@
+      expect(document.querySelector('.unique-input-class')).toBeInTheDocument();
+      expect(document.querySelector('.unique-tag-list-class')).toBeInTheDocument();
+    });
+  });
+
+  describe('id API', () => {
+    it('sets id of container and selected options', () => {
+      render(<MultiSelect {...requiredProps} id="unique-id" value={['Foo']} />);
+
+      expect(document.querySelector('#unique-id')).toBeInTheDocument();
+      expect(getSelectedOption('test label', 'Foo').parentElement).toHaveAttribute('id', 'unique-id-option-Foo');
+    });
+  });
+
+  describe('label API', () => {
+    it('sets label of form control and clear button', () => {
+      render(<MultiSelect {...requiredProps} label="unique label" value={['Foo']} />);
+
+      expect(screen.getByText('unique label')).toBeInTheDocument();
+      expect(getRemoveButton('unique label', 'Foo')).toBeInTheDocument();
+    });
+  });
+
+  describe('listboxChildren API', () => {
+    it('uses render function for listbox children if set', () => {
+      function listboxChildren(options, refs, onSelect) {
+        return (
+          <div>
+            {options.map((option, index) => (
+              <span key={option} ref={refs[index]} onClick={() => { onSelect({ target: refs[index] }); }}>Override {option}</span>
+            ))}
+          </div>
+        );
+      }
+
+      render(<MultiSelect {...requiredProps} listboxChildren={listboxChildren} />);
+
+      openOptions('test label');
+      expect(screen.getByText('Override Foo')).toBeInTheDocument();
+      userEvent.click(screen.getByText('Override Foo'));
+      expect(getSelectedOption('test label', 'Foo')).toBeInTheDocument();
+    });
+  });
+
+  describe('onChange API', () => {
+    it('fires the onChange event when the value changes', () => {
+      let value = [];
+
+      function onChange(event) {
+        value = event.target.value;
+      }
+
+      render(<MultiSelect {...requiredProps} onChange={onChange} />);
+
+      openOptions('test label');
+      userEvent.click(getAvailableOption('test label', 'Foo'));
+      expect(value).toContain('Foo');
+    });
+  });
+
+  describe('options API', () => {
+    it('generates options', () => {
+      render(<MultiSelect {...requiredProps} />);
+
+      openOptions('test label');
+      expect(getAvailableOption('test label', 'Foo')).toBeInTheDocument();
+      expect(getAvailableOption('test label', 'Bar')).toBeInTheDocument();
+    });
+  });
+
+  describe('option objects APIs', () => {
+    it('allows option that are objects, and uses provided getters', () => {
+      render(<MultiSelect
+        {...requiredProps}
+        options={[
+          {
+            id: 'test-1',
+            label: 'Foo',
+          },
+          {
+            id: 'test-2',
+            label: 'Bar',
+          },
+        ]}
+        optionIDGetter={option => option.id}
+        optionLabelGetter={option => option.label}
+      />);
```

Comment:
Seeing this in a test really made me start to think about how I'd like to use this component as an IC outside of Web-inf - I think this is the most common case (`id` or `value` coupled with `label`), I think all strings is less common.

Strings are really only useful for enums, but id/label pair is useful for most things - selecting a user (label is their name / email, id is their .. id) or something like that.

I'd personally say that I think `{ value, label }` is the coupling I've seen the most anecdotally, that's what [angular material uses](https://material.angular.io/components/select/overview), it's what [react-select uses by default](https://react-select.com/home) and I think it's also what [bigmaven typically uses, at least historically](https://github.com/mavenlink/mavenlink/blob/e9b60c29e80a006c9aff0f058c734d83eaa75f7f/frontend/components/inputs/select/select.jsx#L180).

I think providing a familiar API is pretty key to a shared library.

For people who _want_ to use strings, it's obviously very easy for them to do that:

```jsx
const options = ['fart', 'dart', 'cart', 'mart'];
<MultiSelect options={options.map(o => { label: o, value: o })} />
```

I kind of like the idea of providing these getters for when there are edge cases (like `id`/`label`), but one thing that I think makes the _most_ sense is having an opinion here.

If we tell people that the selects take `{ value, label }`, then when the props are handed to the component, they are mapped to this structure, but it only happens once (when the props are passed), instead of a ton of times internally in our component.

```jsx
const users = [{
  id: 2,
  full_name: 'Tom',
}, {
  id: 3,
  full_name: 'Dick',
}, { 
  id: 4,
  full_name: 'Harry',
}];
const options = users.map(user => ({ label: user.full_name, value: user.id }));
<MultiSelect options={options} />
```

I think this is lower friction than telling people to provide a list and two functions on how to fetch the data - just pass us a list in the right format. It can also make our documentation clearer - we don't have to explain the getter functions or how they work, we just say "use `label` and `value` pls". We could even have examples on how to map if that's a concern (which it should not be).

Reply:
Yeah, this is tricky. On the one hand, I think for maximum composability in the simplest use-case, we do want to provide support for string literals. OTOH, if it's functionality we'll literally never use, then why write it in as a possibility?

Further, in the case where an object is the only thing we allow, what's the best default? It seems like it could either be `value` or `id` (`id` makes more sense to me, as that's what we really need, a unique identifier), but even inside your post there doesn't seem to be a real consensus. In our discussion during arch meeting, we couldn't land on a specific format either.

In the case that we do use an object, I think we still want to provide the `getter*` props API. IMO it's much simpler to tell a component how to parse my format in a very explicit way than to have to transform my own objects to pass them in, then do some transformation or key mapping _again_ when I pull the values out.

All this is to say... I can definitely see dropping support for string literals if we think we won't ever use them, and also having a default object format that conforms to `{id: 'unique-identifier', label: 'some-label}`, but I think anything else is going to take some convincing on my end.

Reply:
By way of example for supporting `getter*` props API:
```
const someData = [{account: '1', name: 'test 1'}, {account: '2', name: 'test 2'}...];

// MultiSelect does not support `getter*` props API
<MultiSelect options={someData.map(data => {id: data.account, label: data.name})} ... />

// doing something with value
const dataToSendToBackend = multiSelectRef.current.value.map(val => {account: val.id, name: val.name});
// OR
const dataToSendToAnotherComponent = multiSelectRef.current.value.map(val => someData.find(data => data.account === val.id));

// MultiSelect supports `getter*` props API
<MultiSelect options={someData} optionIDGetter={option => option.account} optionLabelGetter={option => option.name} />

const dataToSendToBackend = multiSelectRef.current.value;
```

Reply:
So a big part of the reason I think `{ value, label }` is used so frequently is because it mimics what a native `<option>` does - 


https://developer.mozilla.org/en-US/docs/Web/HTML/Element/select#basic_select

```html
<select name="choice">
  <option value="first">First Value</option>
  <option value="second" selected>Second Value</option>
  <option value="third">Third Value</option>
</select>
```

`id` is useful for us, sure, but not everything will be ids. What about enum values? The value of an enum isn't necessarily an `id`. It's a _unique identifier_, so if you twisted my arm I'd call it an id, but I would never describe something like `premier_plan` as an id.

I personally think that we can just eschew the getter args _entirely_ in favor of just being opinionated and saying that `options` must always be of the format `{label, value}`, but I could also accept `{label, id}` (it's just not my _favorite way).

I think we can _keep_ the functions, I just don't think they're needed. The way I see it, if we tell consumers of the api that we accept `{label, value}` but we have functions for working around that, there's no reason to provide a default. I'd _rather_ we have the one format and just tell people to make their data structures fit into our world view. 

I don't know if what I said in this comment is any different from my first one - I just kinda feel like every time _I_ have used a `<select>`, I've assumed it accepted an array of `{ label, value }` and been surprised when it accepted something different, primarily because that's what angular, react-select, and native option elements use. I really, genuinely feel that any shared system should [follow the principle of least astonishment](https://en.wikipedia.org/wiki/Principle_of_least_astonishment).

I like that we give folks the freedom to abandon that paradigm when it makes sense (such as having a list of users which are all `{ id, full_name }`), but the way I would get around that is just by mapping over the data before handing it to a `<select>` component.

Reply:
@cbacon91 as an IC using this component, and having a look at my example above, which would you prefer to use?

Reply:
If I'm brutally honest, I have never seen another team use `ref` outside of very specific instances, like when `focus` is relevant, so I would expect that my usage would be more along the lines of 

```jsx
const [selected, setSelected] = useState([]);

const onChange = (data) => {
  setSelected([...selected, data.event.target.ivalue]);
}

<MultiSelect onChange={onChange} />
```

In a case where I'm using callbacks and some parent state, it's 100% sixes. I can use `target.id` or `target.value`, I don't care a ton. The only friction I have at all is the level of configuration I have when building the component initially, and `collection + map`, to me, is easier than `collection + value getter + label getter`.

I'm super happy to be wrong here, I just think if I was still on RPM and was trying to use this select component, this would be my experience.

Reply:
In your example, you've elided the added complexity of data munging by not using `getter*` I was trying to point out. Even if using `onChange`, you still need to munge the data going in and coming out in the case the data you're working with does not already match a prescribed format. Whether or not we use `ref` is not what I was trying to get at.

Reply:
I guess I don't understand.

Regardless, the base of everything I'm trying to say is that what _I would assume_ is that the default structure of `options` is of the format `{ label, value }`. Whether or not there are getters or helpers or something is something, if I'm honest, I don't think I would ever use unless I had to. Added complexity from munging or no, the familiarity of a fairly common `label`/`value` pair would be beneficial to me. I could take or leave the functions. 

What _I_ would prefer the most is that the base assumption (for the default getter / setter functions) is that `options` is an array honoring that structure, but like I said before, it's all sixes. I don't think it _matters_ a ton, I think since the configuration is there, it doesn't matter. I'm personally just a fan of leaving off as much configuration as I can and just rolling with well-reasoned defaults, and I don't think I agree that bare strings are well-reasoned defaults, but _in my opinion_ since the configuration is there, I don't think it's worth discussing a ton more. I just know that 6 months ago, if I saw this component, I know the assumptions I would make, and I feel like I'd be very confused without trudging through documentation.

Reply:
I think we agree that having sane defaults is the way to go, and I'm definitely on board for removing string literals as a possibility for `options`.

I think where we disagree is on whether it's easier to munge a dataset as it goes in/out of the component, or telling the component what it needs to know about the structure of the data going in.

I appreciate the reference to how a native HTML select works! I think that in and of itself is a good enough reason to make the `getter*` default match that format, if we go that route (removing support for string literals). I'm loathe to make any decisions based on what `react-select` does because if we wrote something that complex I would die... But native HTML is great to mimic. 😄

Reply:
Yeah, native HTML is probably the thing to mimic. If we do need a highly-tailored bigmaven interface then we can just make the wrapper that does all the nice to configure. 

I think the power of composition is probably better for us than the ease-of-use of bigmaven interfaces since we can always make the "Project autocompleter" that does the right thing.

Reply:
One thing I'll comment on after working with our `<Select>` component...

I'm a little surprised our [Select component](https://mavenlink.github.io/design-system/master/#/Components/Select) doesn't have this in any fashion. Like, I was trying to make enums work in Custom forms and it pained me just how verbose I had to be 

https://github.com/mavenlink/mavenlink/blob/5d0017c0e974b35a99c5c28892d97016d3ca1e8e/frontend/features/custom-forms/components/field.jsx#L69-L84

I expected to just write something like `<Select options={[1,2,3,4]} />` like we talked about here, or `<Select options={[1,2,3,4].map(r => ({ label: r, value: r }))} />` if it needed an object, but as far as I could tell from the component, neither of those existed - the `select` just runs on children (and I guess `refs`).

I don't think that's worse (or better). On the one hand, it's closer to native by requiring children instead of relying on objects, but on the other hand it's ... just very verbose. It's nice that the option for custom stuff is _there_, but I feel like that's the exception, not the rule.

I kind of like that the multiselect you've made here supports both the verbose / composable case as well as the basic "I have an array of stuff, render as options pls" case. It might be ekeing into `react-select` / god component / 2manyprops territory, but I like the convenience personally

Reply:
Yeah, the `Select` is pretty indisputably more difficult to work with for the reasons you mention, even if it does feel more like a native HTML implementation.

I've also been concerned over the large props surface area `MultiSelect` has at this point. I've done my best to keep it down to as little as possible, but perhaps there is a possibility for even lower-level components that implement less customization capability. The problem then becomes maintaining that simplicity while still allowing for higher-level abstractions to use it compositionally _and_ still be capable of adding the functionality they need.